File: backend/app/metabase.py
from __future__ import annotations

import httpx

from .config import settings


class MetabaseClient:
    def __init__(self) -> None:
        self.base_url = settings.metabase_url.rstrip("/")
        self.api_key = settings.metabase_api_key
        if not self.api_key:
            raise RuntimeError("METABASE_API_KEY is required")

    def _headers(self) -> dict[str, str]:
        return {
            "X-API-KEY": self.api_key,
            "Content-Type": "application/json",
        }

    async def execute_query(self, database_id: int, sql: str) -> list[dict]:
        payload = {
            "database": database_id,
            "type": "native",
            "native": {"query": sql},
        }
        async with httpx.AsyncClient(timeout=120.0) as client:
            response = await client.post(
                f"{self.base_url}/api/dataset",
                headers=self._headers(),
                json=payload,
            )
            response.raise_for_status()
            data = response.json()

        if data.get("status") == "failed":
            raise RuntimeError(data.get("error") or "Metabase query failed")

        cols = [c["name"] for c in data.get("data", {}).get("cols", [])]
        rows = data.get("data", {}).get("rows", [])
        return [dict(zip(cols, row)) for row in rows]

File: frontend/api/metabase.ts
type MetabaseRow = Record<string, unknown>;

export async function executeMetabaseQuery(
  databaseId: number,
  sql: string,
): Promise<MetabaseRow[]> {
  const baseUrl = (process.env.METABASE_URL || "https://metabase.limechat.ai").replace(/\/$/, "");
  const apiKey = process.env.METABASE_API_KEY || "";
  if (!apiKey) {
    throw new Error("METABASE_API_KEY is not configured on the server");
  }

  const response = await fetch(`${baseUrl}/api/dataset`, {
    method: "POST",
    headers: {
      "X-API-KEY": apiKey,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      database: databaseId,
      type: "native",
      native: { query: sql },
    }),
  });

  if (!response.ok) {
    const text = await response.text().catch(() => "");
    throw new Error(`Metabase HTTP ${response.status}: ${text.slice(0, 300)}`);
  }

  const data = (await response.json()) as {
    status?: string;
    error?: string;
    data?: { cols?: Array<{ name: string }>; rows?: unknown[][] };
  };

  if (data.status === "failed") {
    throw new Error(data.error || "Metabase query failed");
  }

  const cols = (data.data?.cols || []).map((c) => c.name);
  const rows = data.data?.rows || [];
  return rows.map((row) => {
    const obj: MetabaseRow = {};
    cols.forEach((name, i) => {
      obj[name] = row[i];
    });
    return obj;
  });
}
